Update CONTRIBUTING.md to reflect current reality

ac000 opened the pull request:
The first patch is a preparatory patch that just re-flows some of the text.

The second patch updates the 'Git Style Guide' section to reflect current reality as it currently gives potential contributors a lot of wrong information.

thresheek commented:
As someone who was stuck by (more than once...) the 72 characters rule, can we maybe also enforce it on the "PR Check" level workflow?  Maybe some other checks too?

(I know this comment is not helpful in this particular set of commits :()

ac000 commented:
> As someone who was stuck by (more than once...) the 72 characters rule, can we maybe also enforce it on the "PR Check" level workflow? Maybe some other checks too?
> 
> (I know this comment is not helpful in this particular set of commits :()

There are of course times when you _shouldn't_ wrap; URLs, command/log output etc...

javorszky commented on CONTRIBUTING.md:
```diff
@@ -47,44 +48,47 @@ the expected behavior that doesn't occur.
 
-- Keep a clean, concise and meaningful `git commit` history on your branch,
-  rebasing locally and squashing before submitting a PR
+- Create atomic commits.  A commit should do just one thing, I.e you
```
Double space, and i.e. is lowercase when in the middle of the sentence, like here, and because it's an abbreviation, the periods are needed after both i and e. See https://www.grammarly.com/blog/know-your-latin-i-e-vs-e-g/
```suggestion
- Create atomic commits. A commit should do just one thing, i.e. you
```

javorszky replied:
Is there a reason for that practice in this file? And I assume other files. I understand that double spaces were important when typewriters were around and I'm missing context on whether that or some similar reason made its way to here.

I also want to gently push back on any explanation that sounds like "we've always done it this way." In those cases I'd like to either figure out what the original reason was, see if that's still relevant, and potentially change it if it no longer serves us. 🙂

ac000 replied:
I've no idea, probably personal writing style.

javorszky commented on CONTRIBUTING.md:
```diff
@@ -47,44 +48,47 @@ the expected behavior that doesn't occur.
-- Keep a clean, concise and meaningful `git commit` history on your branch,
-  rebasing locally and squashing before submitting a PR
+- Create atomic commits.  A commit should do just one thing, I.e you
+  shouldn't mix whitespace changes with actual code changes.  Do the
```
I woulc argue that whitespace changes are also actual code changes. This makes it feel like whitespace changes are somehow lesser than "real" code changes, which isn't something we should allude to. Maybe functional changes?
```suggestion
  shouldn't mix whitespace changes with functional code changes.  Do the
```

ac000 replied:
It of course depends on the language...

In C, whitespace is not significant, whereas in Python it is... seeing as we're a mostly C project, I'd say the above holds, but yes we do have some other language code where whitespace _matters_ e.g Python, so I'll change as suggested.

javorszky commented on CONTRIBUTING.md:
```diff
@@ -47,44 +48,47 @@ the expected behavior that doesn't occur.
-  rebasing locally and squashing before submitting a PR
+- Create atomic commits.  A commit should do just one thing, I.e you
+  shouldn't mix whitespace changes with actual code changes.  Do the
+  whitespace changes in a separate commit first.
```
I don't think the order is important.
```suggestion
  whitespace changes in a separate commit.
```

ac000 replied:
I disagree.

In general you should do any cleanup patches first, e.g as part of preparatory patches for a change.

It's the same as fixing a bunch of coding style issues on some code _before_ you make actual changes to _that_ code.

javorszky replied:
I still think cleanup can come as a later step as long as that's also included. I don't think being prescriptive about having to do cleanup before anything happens is useful. As long as cleanup happens, why should we care in what order that happened?

Sure, individual preferences exist, but those are individual, and we shouldn't push those on others.

ac000 replied:
Except for when you're the one doing the review, I know I'm not the only one that would ask for it to be done that way...

The ordering is somewhat logical. You do preparatory work before _actual_ work.

Let's assume the whitespace changes are trivial and uncontentious, if you _did_ them after the real work and you need to change the real work you may have to also redo the whitespace changes, at the very least these commits will get new id's and just add general noise.

It may also help with things like git-blame(1).

ac000 replied:
`whitespace` changes was just an example, replace that with `refactor`, then tell me that shouldn't come first...

javorszky replied:
It depends. I would treat refactor specifically as its own PR. My argument here is that there are people who want to clean up the whitespace issues / refactor needs first, and then add the new / changed functionality.

Others might want to add the new / changed functionality first, and then tidy it up.

Yet others might want to do cleanup first, add new / changed functionality after, and then tidy it up even more.

I think we should be okay with all kinds of workflows and not prescribe a rigid order.

Everyone should be free to do it in whatever order they see fit as long as the end result passes lint checks, and the individual commits are atomic.

javorszky replied:
> if you did them after the real work and you need to change the real work you may have to also redo the whitespace changes, at the very least these commits will get new id's and just add general noise.

I agree, at the same time we're also asking folks to rebase on their locals and force push anyways, so every time there's a change, most of the commits are going to be totally new anyways. If they also mark the whitespace change commits with a `whitespace` prefix, then as a reviewer we know we can skip that one.

ac000 replied:
> It depends. I would treat refactor specifically as its own PR. My argument here is that there are people who want to clean up the whitespace issues / refactor needs first, and then add the new / changed functionality.
> 
> Others might want to add the new / changed functionality first, and then tidy it up.
> 
> Yet others might want to do cleanup first, add new / changed functionality after, and then tidy it up even more.

If you do things right there shouldn't be any _need_ to _tidy_ things up afterwards. I.e. this is bad form

```
Fixup previous commit
Introduce new feature
```

All it does is add extra noise.

If you do your refactor/cleanup first then it is often the case that your subsequent commits are _simpler_ and _more concise_.

You want commits to be as small and concise as possible. This really helps with reviewing for one thing.

> I think we should be okay with all kinds of workflows and not prescribe a rigid order.
> 
> Everyone should be free to do it in whatever order they see fit as long as the end result passes lint checks, and the individual commits are atomic.

Not really.

At the end of day it's up to the project maintainers how things are done. That is all this commit is doing, updating the contributing document to reflect how things are _currently_ done in Unit.

If you disagree with how the Unit project is currently run then that is a separate conversation.

> > if you did them after the real work and you need to change the real work you may have to also redo the whitespace changes, at the very least these commits will get new id's and just add general noise.
> 
> I agree, at the same time we're also asking folks to rebase on their locals and force push anyways, so every time there's a change, most of the commits are going to be totally new anyways. If they also mark the whitespace

Only from the point of the rebase...

> change commits with a `whitespace` prefix, then as a reviewer we know we can skip that one.

In Unit (like in many other projects) the `prefix` has a specific purpose and saying what _kind_ of commit it is isn't it... also I hope you didn't just say, "_no need to review whitespace changes_"... because they should be reviewed along with everything else.

[ Also just because a patch might be labelled as `whitespace` doesn't mean there won't be _other_ changes in there! ]

javorszky commented on CONTRIBUTING.md:
```diff
@@ -47,44 +48,47 @@ the expected behavior that doesn't occur.
-- In the subject line, use the past tense ("Added feature", not "Add feature");
-  also, use past tense to describe past scenarios, and present tense for
-  current behavior
+- In the subject line, use the imperative mood.  I.e write the subject like
```
Double space, missing period after e. I.e. is at the beginning of the sentence, so capital I is good here.
```suggestion
- In the subject line, use the imperative mood. I.e. write the subject like
```

ac000 replied:
Again following pre-existing style... heh, I've never wrote these things with a trailing period...

javorszky commented on CONTRIBUTING.md:
```diff
@@ -47,44 +48,47 @@ the expected behavior that doesn't occur.
-  also, use past tense to describe past scenarios, and present tense for
-  current behavior
+- In the subject line, use the imperative mood.  I.e write the subject like
+  you're giving git a command, e.g "Free memory before exiting". Do _not_
```
E.g. is also an abbreviation, thus both periods are needed. Punctuation mark in this case goes inside the quotation marks, see https://www.grammarly.com/blog/quotation-marks/

Highlighting the _not_ is a bit harsh in my opinion, we can be friendlier than that. 🙂 
```suggestion
  you're giving git a command, e.g. "Free memory before exiting." Do not
```

ac000 replied:
This is an example of a commit subject and the '.' is _not_ part of the commit subject.

javorszky replied:
Good point, I retract this changeset, thanks!

javorszky commented on CONTRIBUTING.md:
```diff
@@ -47,44 +48,47 @@ the expected behavior that doesn't occur.
-  accordingly
+- Use lowercase subject line prefixes for commits that affect a specific
+  portion of the code; examples include "tests:", "ci:", or "http:", and
+  also individual languages such as "python:" or "php:".  Also if multiple
```
Repetition: also, rewrite, continuing on new line
```suggestion
  also individual languages such as "python:" or "php:". Combine
```

javorszky commented on CONTRIBUTING.md:
```diff
@@ -47,44 +48,47 @@ the expected behavior that doesn't occur.
+- Use lowercase subject line prefixes for commits that affect a specific
+  portion of the code; examples include "tests:", "ci:", or "http:", and
+  also individual languages such as "python:" or "php:".  Also if multiple
+  areas are affected, e.g "auto, perl:"
```
```suggestion
  prefixes if multiple areas are affected, e.g. "auto, perl:"
```

ac000 replied:
Hmm, that suggestion doesn't make sense to me...

ac000 replied:
I've _tweaked_ it anyway...

javorszky replied:
This suggestion goes hand in hand with the Combine one above this. Individually they won't make sense.

javorszky commented on CONTRIBUTING.md:
```diff
@@ -47,44 +48,47 @@ the expected behavior that doesn't occur.
-- Don't rely on command-line commit messages with `-m`; use the editor instead
+- If the commit fixes an open issue then you can use the "Closes:"
+  tag/trailer to reference it and have GitHub automatically close it once
+  it's been merged.  E.g
```
```suggestion
  it's been merged.  E.g.:
```

javorszky reviewed (changes requested):
A few stylistic issues, but otherwise looks good! 🙂

javorszky commented:
An informational question: what's the reason for the 72 char rule?

ac000 commented:
> An informational question: what's the reason for the 72 char rule?

git-log(1) et al, indents the message by four characters. On a standard 80 character wide terminal, wrapping the message after 72 characters gives a four character margin either side. Thus you get a nicely centred message with no ugly wrapping.

This also comes into play when sending and reviewing patches via email...

ac000 commented:
Largely grammatical fixes.
 
````
$ git range-diff 9c6ea795...b26fa9e2
1:  9c6ea795 ! 1:  b26fa9e2 CONTRIBUTING.md: Update the 'Git Style Guide' section
    @@ CONTRIBUTING.md: request issue first to start a discussion about the feature.
      
     -- Keep a clean, concise and meaningful `git commit` history on your branch,
     -  rebasing locally and squashing before submitting a PR
    -+- Create atomic commits.  A commit should do just one thing, I.e you
    -+  shouldn't mix whitespace changes with actual code changes.  Do the
    ++- Create atomic commits.  A commit should do just one thing, i.e. you
    ++  shouldn't mix whitespace changes with functional code changes.  Do the
     +  whitespace changes in a separate commit first.
      
     -- For any user-visible changes, updates, and bugfixes, add a note to
    @@ CONTRIBUTING.md: request issue first to start a discussion about the feature.
     -- In the subject line, use the past tense ("Added feature", not "Add
     -  feature"); also, use past tense to describe past scenarios, and present
     -  tense for current behavior
    -+- In the subject line, use the imperative mood.  I.e write the subject like
    -+  you're giving git a command, e.g "Free memory before exiting". Do _not_
    ++- In the subject line, use the imperative mood.  I.e. write the subject like
    ++  you're giving git a command, e.g. "Free memory before exiting". Do not
     +  terminate the subject with a `.`
      
     -- Limit the subject line to 67 characters, and the rest of the commit message
    @@ CONTRIBUTING.md: request issue first to start a discussion about the feature.
     -  accordingly
     +- Use lowercase subject line prefixes for commits that affect a specific
     +  portion of the code; examples include "tests:", "ci:", or "http:", and
    -+  also individual languages such as "python:" or "php:".  Also if multiple
    -+  areas are affected, e.g "auto, perl:"
    ++  also individual languages such as "python:" or "php:".  If multiple areas
    ++  are affected you can specify multiple prefixes, e.g. "auto, perl:"
      
     -- Don't rely on command-line commit messages with `-m`; use the editor instead
     +- If the commit fixes an open issue then you can use the "Closes:"
     +  tag/trailer to reference it and have GitHub automatically close it once
    -+  it's been merged.  E.g
    ++  it's been merged.  E.g.:
     +
     +  `Closes: https://github.com/nginx/unit/issues/9999`
     +
```

ac000 commented:
Change the no mixed changes example to 'refactoring'

```
$ git range-diff b26fa9e2...db59b529
1:  b26fa9e2 ! 1:  db59b529 CONTRIBUTING.md: Update the 'Git Style Guide' section
    @@ CONTRIBUTING.md: request issue first to start a discussion about the feature.
     -- Keep a clean, concise and meaningful `git commit` history on your branch,
     -  rebasing locally and squashing before submitting a PR
     +- Create atomic commits.  A commit should do just one thing, i.e. you
    -+  shouldn't mix whitespace changes with functional code changes.  Do the
    -+  whitespace changes in a separate commit first.
    ++  shouldn't mix refactoring with functional code changes.  Do the
    ++  refactoring in one or more commits first.
      
     -- For any user-visible changes, updates, and bugfixes, add a note to
     -  `docs/changes.xml` under the section for the upcoming release, using
```

javorszky commented:
> > An informational question: what's the reason for the 72 char rule?
> 
> git-log(1) et al, indents the message by four characters. On a standard 80 character wide terminal, wrapping the message after 72 characters gives a four character margin either side. Thus you get a nicely centred message with no ugly wrapping.
> 
> This also comes into play when sending and reviewing patches via email...

Neat, thank you! Today I learned.

javorszky commented on CONTRIBUTING.md:
```diff
@@ -47,44 +48,47 @@ the expected behavior that doesn't occur.
+  `Closes: https://github.com/nginx/unit/issues/9999`
+
+That should go at the end of the commit message, separated by a blank line,
+along with any other tags.
```
Not a blocker. I understand this belongs with the previous point about the closes tag. Does this need to be indented to render as being part of the bullet point?

ac000 replied:
Hmm, I think it's OK the way it is, it's indented at the same level of the bullet point it's part of and if there were more bullet points after it I think it'd still be clear what it belongs to...

javorszky replied:
I might be missing something. When I go to view this file at this commit (https://github.com/nginx/unit/blob/db59b52964e209fb1d954a1a93bb1481ccca3f8e/CONTRIBUTING.md), this is what it renders as:

![image](https://github.com/user-attachments/assets/1a05c032-7820-4c53-bf7c-9b85e590b128)

It looks like that lost line isn't part of the last bullet point.

ac000 replied:
Yes, that's what I see...

ac000 replied:
Oh, the _last_ line, yes that should perhaps be indented...

javorszky reviewed (approved):
One tiny non-blocker question, otherwise good to go! Thank you 🙂

ac000 commented:
Fix some indentation

```
$ git range-diff db59b529...6d6ea136
1:  db59b529 ! 1:  6d6ea136 CONTRIBUTING.md: Update the 'Git Style Guide' section
    @@ CONTRIBUTING.md: request issue first to start a discussion about the feature.
     +
     +  `Closes: https://github.com/nginx/unit/issues/9999`
     +
    -+That should go at the end of the commit message, separated by a blank line,
    -+along with any other tags.
    ++  That should go at the end of the commit message, separated by a blank line,
    ++  along with any other tags.
      
```

ac000 commented:
Rebase with master

```
$ git range-diff 6d6ea136...20224279
 -:  -------- >  1:  6ca27c32 contrib: make sha512sum check compatible with FreeBSD 14+
 -:  -------- >  2:  081e5115 status: Constify a bunch of local variables
 -:  -------- >  3:  c8d70c3f status: Use a variable to represent the status member index
 -:  -------- >  4:  55041ef9 Flow the language module name into nxt_app_lang_module_t
 -:  -------- >  5:  707f4ef8 status: Show list of loaded language modules
 -:  -------- >  6:  f4ba4b55 tests: Fix `/status' endpoint tests for new 'modules' section
 -:  -------- >  7:  ebb10b0a Fix a comment typo for 'Memory-only buffers' in src/nxt_buf.h
 -:  -------- >  8:  1c607662 status: Add a missing check for potential NULL
 -:  -------- >  9:  58fdff54 fuzzing: added cifuzz workflow
 -:  -------- > 10:  fcbaf8f3 fuzzing: fix harness bugs
 -:  -------- > 11:  61c13ade fuzzing: update directory path in README and build-fuzz.sh
 -:  -------- > 12:  90542dbd ci: cifuzz: Bump github/codeql-action from 2 to 3
 -:  -------- > 13:  57a75ea0 build(deps): bump openssl from 0.10.64 to 0.10.66 in /tools/unitctl
 -:  -------- > 14:  b892e994 tools/unitctl: update readme
 -:  -------- > 15:  1b484303 tools/unitctl: update readme
 -:  -------- > 16:  e743b6ce tools/unitctl: remove (default) from option text
 -:  -------- > 17:  2a243740 tools/unitctl: make json-pretty default output fmt
 -:  -------- > 18:  43faf99d tools/unitctl: reword freeform message for output
 -:  -------- > 19:  a91b961d tools/unitctl: make application directory configurable
 -:  -------- > 20:  e56c4ede fuzzing: code cleanup
 -:  -------- > 21:  900d25c3 fuzzing: fixed harness bug
 -:  -------- > 22:  bc49274d fuzzing: updated JSON target
 -:  -------- > 23:  3667c3e2 fuzzing: added new basic targets
 -:  -------- > 24:  06c4ea1f Add a basic .editorconfig file
 1:  14a5e66b = 25:  0f313e2b CONTRIBUTING.md: Re-flow text
 2:  6d6ea136 = 26:  20224279 CONTRIBUTING.md: Update the 'Git Style Guide' section
```
